Sources/Basic/OrderedDictionary.swift

```diff
@@ -0,0 +1,114 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2018 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+/// A generic collection to store key-value pairs in the order they were inserted in.
+///
+/// This is modelled after the stdlib's Dictionary.
+public struct OrderedDictionary<Key: Hashable, Value> {
+
+    /// The element type of a dictionary: a tuple containing an individual
+    /// key-value pair.
+    public typealias Element = (key: Key, value: Value)
+
+    /// The underlying storage for the OrderedDictionary.
+    fileprivate var array: [Key]
+    fileprivate var dict: [Key: Value]
+
+    /// Create an empty OrderedDictionary object.
+    public init() {
```

Comment:
A dictionaryLiteral initializer would be nice

Reply:
Yep, there is an extension for that.